src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2510,9 +2547,15 @@ InitializeWorkerNodeCache(void)
 	int hashFlags = 0;
 	long maxTableSize = (long) MaxWorkerNodesTracked;
 	bool includeNodesFromOtherClusters = false;
+	int workerNodeIndex = 0;
 
 	InitializeCaches();
 
+	if (WorkerNodeArray != NULL)
+	{
+		pfree(WorkerNodeArray);
+	}
```

Comment:
It makes me nervous to call `pfree` and have such a large gap before `Alloc` is called.

- This should also set `WorkerNodeArray` to `NULL`, just in case this function accidentally calls something which uses `WorkerNodeArray`.
- I can't think of a reason why free-ing early would cause problems but this same function seems to go out of it's way to not destroy `oldWorkerNodeHash` until the very end, is that vestigial?

Reply:
Refactored it to do all the re-initialisation at the end.